File: class/Recorder/DataRecorder.m
classdef DataRecorder < DataRecorderInitializer
	% DATARECORDER
	%
	properties
		recordId;
        file_count;
	    
		manager;
	end
	
	methods
		function obj = DataRecorder(varargin)
			obj.set(varargin);
        end
        
		function set_inner(obj, argvnum, argvstr, argvdata)
			if nargin < 2 || argvnum == 0
				obj.set_inner@DataRecorderInitializer(0);
            else
				obj.set_inner@DataRecorderInitializer(argvnum, argvstr, argvdata);
				for i = 1:argvnum
					switch argvstr{i}
						case {'mode'}
							obj.mode = argvdata{i};
                        case {'name', 'recordName', 'n'}
                            obj.recordName = argvdata{i};
						otherwise
					end
				end
            end
			obj.reset();
		end

		function setId(obj)
   			obj.recordId = IdManager.getRecorderCount(obj.manager);
        end
		
		function reset(obj)
            if isempty(obj.recordId)
    			obj.setId();
                obj.file_count = 0;
            end
            rm = RecordManager.getObject();
            if isempty(rm)
                warning('RecordManager is not defined.');
                evalin('base', 'recordManager_ = RecordManager();');
                rm = RecordManager.getObject();
            end
            obj.manager = rm;
            obj.isValid = true;
            if strcmp(obj.flag, 'registerObserver')
                obj.registerObserver();
            end
        end
        
        function registerObserver(obj)
            if isempty(obj.observer) || ~obj.observer.isValid
                warning('Observer is not ready.');
                return;
            end
            if isempty(obj.observer.recorder) || obj.observer.recorder.recordId ~= obj.recordId
                obj.observer.set('recorder', obj);
            end
            obj.flag = '';
        end
    end
end

File: class/Recorder/DataRecorder.cpp
#include "DataRecorder.h"

using namespace RNNSimulator;
#ifndef _DATARECORDER_CPP_
#define _DATARECORDER_CPP_
int DataRecorder::NEXTRECID = 0;
#endif

DataRecorder::DataRecorder() : DataRecorder("") {};
DataRecorder::DataRecorder(std::string recordName_) {
	if (recordName_.length() > 0) {
		recordName = recordName_;
	}
	reset();
	isValid = true;
}
DataRecorder::~DataRecorder() {}

void DataRecorder::reset() {
	if (simId < 0) setId();
}
void DataRecorder::setId() {
	if (simId >= 0) return;
	// read simulation id from idfile

	// set recordId and increment NEXTRECID
	if (recordId < 0) {
		recordId = NEXTRECID;
		NEXTRECID += 1;
	}
}
void DataRecorder::recordObserver(Observer* observer_) {
	observer = observer_;
}

int DataRecorder::set_inner(int varargnum, va_list argv) {
	std::string str;
	for (int i = 0;i < varargnum;i++) {
		str = va_arg(argv, char*);
		if (str == "observer" || str == "obs" || str == "o") {
			recordObserver(va_arg(argv, Observer*));
		}
		else if (str == "recordName" || str == "name" || str == "n") {
			recordName = va_arg(argv, char*);
		}
	}
	reset();
	return varargnum;
}